File: backend/app/models/agent.py
"""Models for AI Agent system with tool use and human-in-the-loop approval."""

from datetime import datetime
from enum import Enum
from typing import Optional
from uuid import UUID, uuid4

from pydantic import BaseModel as PydanticBaseModel
from sqlalchemy import Column
from sqlalchemy.dialects.postgresql import JSONB
from sqlmodel import Field, SQLModel


class AgentTaskStatus(str, Enum):
    """Status of an agent task."""
    PENDING = "pending"  # Waiting to be processed
    RUNNING = "running"  # Agent is working
    AWAITING_APPROVAL = "awaiting_approval"  # Needs human approval for action
    APPROVED = "approved"  # Action approved, executing
    REJECTED = "rejected"  # Action rejected by human
    COMPLETED = "completed"  # Task finished successfully
    FAILED = "failed"  # Task failed with error
    CANCELLED = "cancelled"  # Task cancelled by user


class AgentActionType(str, Enum):
    """Types of agent actions that may require approval."""
    READ_ONLY = "read_only"  # Safe, no approval needed
    CREATE = "create"  # Creating new records
    UPDATE = "update"  # Modifying existing records
    DELETE = "delete"  # Removing records
    SEND = "send"  # Sending emails/messages
    BULK = "bulk"  # Bulk operations


class AgentTask(SQLModel, table=True):
    """Database model for agent tasks."""

    __tablename__ = "agent_tasks"

    id: UUID = Field(default_factory=uuid4, primary_key=True)

    # Task details
    prompt: str  # The original user request
    goal: str  # What the agent is trying to accomplish
    status: AgentTaskStatus = Field(default=AgentTaskStatus.PENDING, index=True)

    # Execution tracking
    steps_completed: int = Field(default=0)
    max_steps: int = Field(default=10)
    current_step: Optional[str] = None

    # Results
    result: Optional[str] = None
    error: Optional[str] = None

    # Action history (for audit trail)
    action_history: list = Field(default_factory=list, sa_column=Column(JSONB))

    # Pending action requiring approval
    pending_action: Optional[dict] = Field(default=None, sa_column=Column(JSONB))

    # Timestamps
    created_at: datetime = Field(default_factory=datetime.utcnow)
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None

    # AI metadata
    model_version: str = Field(default="llama3.1")


class AgentAction(SQLModel, table=True):
    """Database model for individual agent actions (audit trail)."""

    __tablename__ = "agent_actions"

    id: UUID = Field(default_factory=uuid4, primary_key=True)
    task_id: UUID = Field(foreign_key="agent_tasks.id", index=True)

    # Action details
    action_type: AgentActionType
    tool_name: str
    tool_input: dict = Field(default_factory=dict, sa_column=Column(JSONB))
    tool_output: Optional[dict] = Field(default=None, sa_column=Column(JSONB))

    # Approval tracking
    requires_approval: bool = Field(default=False)
    approved: Optional[bool] = None
    approved_at: Optional[datetime] = None
    rejection_reason: Optional[str] = None

    # Execution
    executed: bool = Field(default=False)
    execution_error: Optional[str] = None

    # Timestamps
    created_at: datetime = Field(default_factory=datetime.utcnow)


# Response models
class AgentTaskResponse(PydanticBaseModel):
    """API response for an agent task."""
    id: UUID
    prompt: str
    goal: str
    status: AgentTaskStatus
    steps_completed: int
    max_steps: int
    current_step: Optional[str]
    result: Optional[str]
    error: Optional[str]
    pending_action: Optional[dict]
    created_at: datetime
    started_at: Optional[datetime]
    completed_at: Optional[datetime]

    class Config:
        from_attributes = True


class AgentActionResponse(PydanticBaseModel):
    """API response for an agent action."""
    id: UUID
    task_id: UUID
    action_type: AgentActionType
    tool_name: str
    tool_input: dict
    tool_output: Optional[dict]
    requires_approval: bool
    approved: Optional[bool]
    executed: bool
    execution_error: Optional[str]
    created_at: datetime

    class Config:
        from_attributes = True


class CreateAgentTaskRequest(PydanticBaseModel):
    """Request to create a new agent task."""
    prompt: str
    max_steps: int = 10


class ApproveActionRequest(PydanticBaseModel):
    """Request to approve or reject a pending action."""
    approved: bool
    rejection_reason: Optional[str] = None

File: frontend/src/types/agent.ts
/**
 * AI Agent Types - Autonomous task execution with tool use
 */

// Enums
export type AgentTaskStatus =
  | "pending"
  | "running"
  | "awaiting_approval"
  | "approved"
  | "rejected"
  | "completed"
  | "failed"
  | "cancelled";

export type AgentActionType =
  | "read_only"
  | "create"
  | "update"
  | "delete"
  | "send"
  | "bulk";

// Main types
export interface AgentTask {
  id: string;
  prompt: string;
  goal: string;
  status: AgentTaskStatus;
  steps_completed: number;
  max_steps: number;
  current_step: string | null;
  result: string | null;
  error: string | null;
  pending_action: PendingAction | null;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
}

export interface PendingAction {
  tool_call_id: string;
  tool_name: string;
  tool_input: Record<string, unknown>;
  action_type: AgentActionType;
}

export interface AgentAction {
  id: string;
  task_id: string;
  action_type: AgentActionType;
  tool_name: string;
  tool_input: Record<string, unknown>;
  tool_output: Record<string, unknown> | null;
  requires_approval: boolean;
  approved: boolean | null;
  executed: boolean;
  execution_error: string | null;
  created_at: string;
}

export interface AgentTool {
  name: string;
  description: string;
  parameters: {
    type: string;
    properties: Record<string, unknown>;
    required: string[];
  };
}

export interface AgentStatus {
  available: boolean;
  enabled: boolean;
  ollama_status: string;
  model: string;
  require_approval_for_writes: boolean;
  available_tools: number;
}

// Request types
export interface CreateAgentTaskRequest {
  prompt: string;
  max_steps?: number;
}

export interface ApproveActionRequest {
  approved: boolean;
  rejection_reason?: string;
}

// Helper functions
export function getStatusColor(status: AgentTaskStatus): string {
  switch (status) {
    case "completed":
      return "text-green-600";
    case "running":
      return "text-blue-600";
    case "awaiting_approval":
      return "text-yellow-600";
    case "failed":
    case "rejected":
    case "cancelled":
      return "text-red-600";
    default:
      return "text-gray-600";
  }
}

export function getStatusBgColor(status: AgentTaskStatus): string {
  switch (status) {
    case "completed":
      return "bg-green-100 text-green-800";
    case "running":
    case "approved":
      return "bg-blue-100 text-blue-800";
    case "awaiting_approval":
      return "bg-yellow-100 text-yellow-800";
    case "failed":
    case "rejected":
    case "cancelled":
      return "bg-red-100 text-red-800";
    default:
      return "bg-gray-100 text-gray-800";
  }
}

export function getActionTypeLabel(type: AgentActionType): string {
  switch (type) {
    case "read_only":
      return "Read";
    case "create":
      return "Create";
    case "update":
      return "Update";
    case "delete":
      return "Delete";
    case "send":
      return "Send";
    case "bulk":
      return "Bulk Operation";
    default:
      return "Action";
  }
}

export function getActionTypeBgColor(type: AgentActionType): string {
  switch (type) {
    case "read_only":
      return "bg-gray-100 text-gray-800";
    case "create":
      return "bg-green-100 text-green-800";
    case "update":
      return "bg-blue-100 text-blue-800";
    case "delete":
      return "bg-red-100 text-red-800";
    case "send":
      return "bg-purple-100 text-purple-800";
    case "bulk":
      return "bg-orange-100 text-orange-800";
    default:
      return "bg-gray-100 text-gray-800";
  }
}

export function formatToolName(name: string): string {
  return name
    .split("_")
    .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
    .join(" ");
}
